cmake/BDWGC.cmake

```diff
@@ -0,0 +1,74 @@
+macro(p4c_obtain_bdwgc)
+  include(CheckSymbolExists)
+
+  option(
+    P4C_USE_PREINSTALLED_BDWGC
+    "Look for a preinstalled version of BDWGC in the system instead of installing a prebuilt binary using FetchContent."
+    OFF
+  )
+  set(LIBGC_LIBRARIES gc gccpp cord)
+  set(HAVE_LIBGC 1)
+
+  if(P4C_USE_PREINSTALLED_BDWGC)
+    OPTION (ENABLE_MULTITHREAD "Use multithreading" OFF)
+    find_package(LibGc 7.2.0 REQUIRED)
+    check_function_exists (GC_print_stats HAVE_GC_PRINT_STATS)
+  else()
+    # Print out download state while setting up BDWGC.
+    set(FETCHCONTENT_QUIET_PREV ${FETCHCONTENT_QUIET})
+    set(BUILD_SHARED_LIBS_PREV ${BUILD_SHARED_LIBS})
+    set(FETCHCONTENT_QUIET OFF)
+    set(BUILD_SHARED_LIBS OFF CACHE BOOL "")
+
+    # Turn on C++ support.
+    set(enable_cplusplus ON CACHE BOOL "C++ support")
+    set(install_headers OFF CACHE BOOL "Install header and pkg-config metadata files.")
+    set(enable_docs ${ENABLE_DOCS} CACHE BOOL "Build and install documentation.")
+    # We may have to compile very large programs which allocate a lot on the heap.
+    set(enable_large_config ON CACHE BOOL "Optimize for large heap or root set.")
+    set(without_libatomic_ops OFF CACHE BOOL "Use atomic_ops.h in libatomic_ops/src")
+    # Redirect all malloc calls in the program.
+    set(enable_redirect_malloc OFF CACHE BOOL "Redirect malloc and friends to GC routines.")
+    # Try to enable thread-local-storage for better performance.
+    set(enable_thread_local_alloc ON CACHE BOOL "Turn on thread-local allocation optimization")
+
+    fetchcontent_declare(
+      bdwgc
+      GIT_REPOSITORY https://github.com/ivmai/bdwgc.git
+      GIT_TAG e340b2e869e02718de9c9d7fa440ef4b35785388 # 8.2.6
```

Comment:
Is there an easy way for someone to override this if they want to use their own repo/copy of bdwgc instead of the public one?  This would apply to other things we're pulling with fetchcontent (protobuf, abseil, etc) -- it would be nice if there was one place all the references (repo URLs and tags) were collected rather than being scattered across different CMakefiles for each dependency.

Reply:
Are you looking for a central CMake file that contains all the Github tags of dependencies we pull? We could define that and use it for all core FetchContent dependencies. So far that would be Abseil, Protobuf, Gtest, and BDWGC. 

I think where this gets complicated is with modularity. For example, the GC dependency is only pulled in when `ENABLE_GC` is toggled. `Protobuf` is also only meant to be installed when we enable the control plane.

Reply:
I think a central dependency repository might be tricky since every back end manages their own dependencies. There is [string(JSON...)](https://cmake.org/cmake/help/latest/command/string.html#json) which we could use to parse a JSON file with all the necessary information on dependencies. That file could also be user-provided.

We could also just create a helper `fetchcontent_declare_with_option` which adds CMake options to set the URL, URL_HASH and GIT TAG for each `fetchcontent_declare` invocation.